crates/shared/src/event_handling.rs

```diff
@@ -69,118 +60,169 @@ where
-            } else {
-                self.store.append_events(unwrapped_events).await?;
-            };
-        }
-        // The `chunks` adaptor does not return an empty chunk if the stream was completely empty.
-        // However we do want to delete old events in this case as a rerorg might have removed
-        // events without adding new ones.
-        if !have_deleted_old_events {
-            self.store.replace_events(Vec::new(), range.clone()).await?;
+        let replacement_blocks = self.event_block_range().await?;
+        tracing::debug!(
+            "replacement_blocks before filtering: {:?}",
+            replacement_blocks
+        );
+        let (replacement_blocks, events) = self.past_events(&replacement_blocks).await;
+        track_block_range(&format!("range_{}", replacement_blocks.len()));
+        tracing::debug!(
+            "replacement_blocks after filtering: {:?}, events number: {}",
+            replacement_blocks,
+            events.len()
+        );
+        if replacement_blocks.is_empty() {
+            return Ok(());
         }
-        self.last_handled_block = Some(range.end().to_u64());
+
+        let range = RangeInclusive::new(
+            replacement_blocks.first().unwrap().0,
+            replacement_blocks.last().unwrap().0,
+        );
+        self.store.replace_events(events, range.clone()).await?;
+
+        // delete forked blocks
+        self.last_handled_blocks
+            .retain(|block| block.0 < replacement_blocks.first().unwrap().0);
+        // append new canonical blocks
+        self.last_handled_blocks
+            .extend(replacement_blocks.into_iter());
+        // cap number of blocks to MAX_REORG_BLOCK_COUNT
+        let start_index = self
+            .last_handled_blocks
+            .len()
+            .saturating_sub(MAX_REORG_BLOCK_COUNT as usize);
+        self.last_handled_blocks = self.last_handled_blocks[start_index..].to_vec();
+
         Ok(())
     }
 
     async fn past_events(
         &self,
-        block_range: &RangeInclusive<BlockNumber>,
-    ) -> Result<impl Stream<Item = Result<EthcontractEvent<C::Event>>>, ExecutionError> {
-        Ok(self
-            .contract
-            .get_events()
-            .from_block((*block_range.start()).block_number())
-            .to_block((*block_range.end()).block_number())
-            .block_page_size(500)
-            .query_paginated()
-            .await?
-            .map_err(Error::from))
+        blocks: &[BlockNumberHash],
+    ) -> (Vec<BlockNumberHash>, Vec<EthcontractEvent<C::Event>>) {
+        let (mut blocks_filtered, mut events) = (vec![], vec![]);
+        for block in blocks {
+            match self.contract.get_events().block_hash(block.1).query().await {
+                Ok(e) => {
+                    blocks_filtered.push(*block);
+                    events.extend(e.into_iter());
+                }
+                Err(_) => return (blocks_filtered, events),
+            }
+        }
```

Comment:
Can we turn this into a batched request?

Reply:
Do you maybe have a suggestion how to do this through `EventRetrieving` trait? After some thinking I am not sure how to do this.
Also not sure if we need to batch this since it will be split again through loadbalanced node. Maybe parallel call is the easier solution here?

Reply:
Eventually ended up with parallelized call. Let me know if you think this is ok.